### PR TITLE
Add recursive validation check for configs

### PR DESCRIPTION
**Pull request by bogdandrutu**
Fixes https://github.com/open-telemetry/opentelemetry-collector/issues/4584

### CODE REVIEW

**Comment by dmitryax** [comment]
@bogdandrutu can you please rebase it?

File: component/config.go [hunk 1]
```diff
@@ -33,12 +37,43 @@ type ConfigValidator interface {
...
+	return validate(reflect.ValueOf(cfg))
+}
+
+func validate(t reflect.Value) error {
+	// Validate the value itself.
+	var errs error
+	// If implements ConfigValidator then call Validate.
+	if validator, ok := t.Interface().(ConfigValidator); ok {
+		errs = multierr.Append(errs, validator.Validate())
+	}
```
Comment (dmitryax) [inline component/config.go:49]: Maybe we should validate nested fields first? I can imagine that some validation logic relying on already validated nested fields.

Comment (dmitryax) [reply component/config.go:49]: Actually nvm. Since we don't return and combine all the errors, it only affects the order of the errors.

**Review by dmitryax — COMMENTED** [review]
I like the idea. It would be great to see how it works on some unit tests

**Comment by bogdandrutu** [comment]
> I like the idea. It would be great to see how it works on some unit tests 

Maybe you have an idea how to fix the "type errType string" tests. With reflect I see that as a "string" not as a struct...

**Comment by bogdandrutu** [comment]
> It would be great to see how it works on some unit tests

Lots were added.

File: component/config.go [hunk 2]
```diff
@@ -33,12 +37,43 @@ type ConfigValidator interface {
...
+		t = t.Elem()
+	}
+	switch t.Kind() {
+	case reflect.Struct:
+		// Reflect on the pointed data and check each of its fields.
+		for i := 0; i < t.NumField(); i++ {
+			if !t.Type().Field(i).IsExported() {
+				continue
+			}
+			errs = multierr.Append(errs, validate(t.Field(i)))
```
[the remark this answers is not in the document]
Comment (bogdandrutu) [reply component/config.go]: Not sure, I don't expect users to know about the "structs", especially if we embed/squash. So I expect the error messages to be good enough though.

[the remark this answers is not in the document]
Comment (dmitryax) [reply component/config.go]: That makes sense, but what if the same struct is used in as value of several fields, e.g. `MetricSettings` in this struct https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/receiver/hostmetricsreceiver/internal/scraper/processscraper/internal/metadata/generated_metrics.go#L40? If potential `MetricSettings.Validate()` returns an error, user don't know which metric is configured incorrectly.

[the remark this answers is not in the document]
Comment (bogdandrutu) [reply component/config.go]: Do user know about "MetricSettings"?

[the remark this answers is not in the document]
Comment (dmitryax) [reply component/config.go]: Right, calculation of the mapstruct tree would be required. I'm not saying it's needed in this PR, just to plan for possible future improvements

**Review by dmitryax — APPROVED** [review]
LGTM

**Pull request merged**